src/System.Linq/src/System/Linq/Take.SizeOpt.cs

```diff
@@ -3,21 +3,59 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
+using System.Diagnostics;
 
 namespace System.Linq
 {
     public static partial class Enumerable
     {
         private static IEnumerable<TSource> TakeIterator<TSource>(IEnumerable<TSource> source, int count)
         {
-            if (count > 0)
+            Debug.Assert(count > 0);
+
+            foreach (TSource element in source)
+            {
+                yield return element;
+                if (--count == 0) break;
+            }
+        }
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(count > 0);
+
+            if (source is ICollection<TSource> collection)
+            {
+                if (collection.Count > count)
+                {
+                    return collection.Skip(collection.Count - count);
+                }
+                else if (collection.Count > 0)
+                {
+                    return TakeIterator<TSource>(collection, collection.Count);
+                }
+                else
+                {
+                    return Empty<TSource>();
+                }
+            }
+            else if (source is IReadOnlyCollection<TSource> roCollection)
             {
-                foreach (TSource element in source)
+                if (roCollection.Count > count)
                 {
-                    yield return element;
-                    if (--count == 0) break;
+                    return roCollection.Skip(roCollection.Count - count);
+                }
+                else if (roCollection.Count > 0)
+                {
+                    return TakeIterator<TSource>(roCollection, roCollection.Count);
+                }
+                else
+                {
+                    return Empty<TSource>();
                 }
             }
+
+            return TakeLastRegularIterator<TSource>(source, count);
```

Comment:
The point of the .SizeOpt.cs is to minimize overhead associated with generic instantations.  Unless we're measuring the size impact of this with AOT compilation, let's avoid adding such throughput optimizations to this implementation.

Reply:
So what should be done here - both TakeIterator<T> and mentioned TakeLastRegularIterator<T> should be inlined in TakeLastIterator method?

Reply:
I was commenting more on all of the special cases above this.  I suggest you just make it:
```C#
private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count) =>
    TakeLastRegulatorIterator(source, count);
```
That unfortunately still adds an extra generic method, but it avoids having to duplicate all of the code in TakeLastRegularIterator.

Also, naming-wise, the method you now have named TakeLastIterator isn't actually an iterator.  I suggest you rename what you currently have named TakeLastRegulatorIterator back to being named TakeLastIterator (it actually is an iterator), and then rename what you currently have named TakeLastIterator to be something like TakeLastEnumerableFactory, since it's really just a factory that determines which enumerable implementation to return.